src/vsr/replica.zig

```diff
@@ -1515,20 +1515,17 @@ pub fn Replica(
             }
 
             {
-                const slot_clean = !self.journal.dirty.bit(slot);
-                const slot_faulty = self.journal.faulty.bit(slot);
+                // We may have guaranteed the prepare but our copy is faulty (not safe to nack).
+                if (self.journal.faulty.bit(slot)) return;
                 if (self.journal.header_with_op_and_checksum(message.header.op, checksum)) |_| {
-                    // Either:
-                    // * We have guaranteed the prepare and our copy is clean (not safe to nack).
-                    // * We have guaranteed the prepare but our copy is faulty (not safe to nack —
-                    //   we don't know what op we prepared in this slot).
-                    if (slot_clean or slot_faulty) return;
-                    // We know of the prepare but we have yet to write or guarantee it (safe to nack).
-                    // Continue through below...
+                    if (self.journal.dirty.bit(slot)) {
```

Comment:
I just realized that we may need to also add `prepare_checksums[slot.index] != checksum` as an additional condition here, so that journal recovery from case `@I` (where we do have the prepare but mark it dirty) does not cause us to NACK.

Reply:
If my understanding is correct, would you please add this condition?

Reply:
I think `prepare_checksums[slot.index] != checksum` is already guaranteed at this point; it was checked earlier: https://github.com/coilhq/tigerbeetle/blob/f3134c4774a9b98cfac21e5580540c006c652212/src/vsr/replica.zig#L1476-L1481.

Reply:
Ah thanks, it would be good to add as an assert then on line 1548 just before we send the nack, since 1476 is an optimization, and someone might remove it later without realizing it's also required for correctness.

Reply:
Great point; thank you.